src/digraph/Dominator_Tree.jl

```diff
@@ -0,0 +1,330 @@
+function dominator_Tree(g::AG, source) where {T, AG<:DiGraph{T}}
+
+
+
+  parent=zeros(T,nv(g)) #the parents of every node in the dfs teree
+
+
+  #= the semi dominator is the first node on the path between the root and the node of interest,
+  such that  there is at least two disjoint paths between it and the node of interest ,
+  if there is not such a node we will consider the parent of the node of interst to be the semi dominator of it
+   there is a duality between the semi dominator and the  immediate dominator
+   min(preorderof(semi_dominator(v)))>=max(preorder(dominator(v))
+   for more details see the paper
+   the key theorm is
+   Let w != the source, and let u be a vertex for which semi_dominator(u) is minimum
+   among vertices between w and its semi dominator.
+   Then the immediate dominator is
+   semi_dominator(w) if semi_dominator(w) = semi_dominator(u),
+   immidiate_dominator(w) =  immidiate_dominator(u) otherwise.
+  =#
+  semi=zeros(T,nv(g))
+  #==#
+
+
+  function accumfunc(a1,a2)
+    if semi[a1]<=semi[a2]
+      return true
+    else
+      return false
+    end
+  end
```

Comment:
Why is this an included function? Also, you don't need to do all of this:
```
accumfunc(semi, a1, a2) = semi[a1] <= semi[a2]
```
which means you probably don't need it at all.

Reply:
the semi array is needed in  produce_eval_link function and its value gets modified in the dominator_Tree function , i could have just passed semi array to  produce_eval_link function,
but i don’t think passing pointers to functions is a good programming habit, as the same value can be modified from different functions .